### ISSUE TITLE
Support for DaemonSet eviction when draining nodes

### ISSUE THREAD

**Issue by ailurarctos** [kubernetes-sigs/cluster-api#6158]
(I'm not sure if this feature request is large enough to require the CAEP process. If it is please let me know.)

**User Story**

As a user I would like to some mechanism to have my DaemonSet pods gracefully terminated when draining nodes for deletion so that those pods can complete their shutdown process.

**Detailed Description**

Currently Cluster API uses the standard kubectl drain ignoring all DaemonSets ([link](https://github.com/kubernetes-sigs/cluster-api/blob/d3f906087fa13686a62dd78e07d358d57529e73f/internal/controllers/machine/machine_controller.go#L519)). I would like some way to have my DaemonSet pods also gracefully terminated as part of the node deletion process.

**Anything else you would like to add:**

While investigating whether this is currently possible I saw that Cluster Autoscaler provides a [mechanism](https://github.com/kubernetes/autoscaler/blob/0d9fb78c9258cacab29b8a6cd37caf60edfc0c2e/cluster-autoscaler/FAQ.md#how-can-i-enabledisable-eviction-for-a-specific-daemonset) to control DaemonSet draining. I'm planning to make use of this in the interim but it would be nice to also have the draining happen for when nodes are not drained by Cluster Autoscaler (e.g. for cluster upgrades, etc.).

I also looked into the [graceful node shutdown](https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown) feature but in my case the pod drain time is quite long (could be 30 minutes or longer) and I'm not sure the feature would work for such long termination times, especially in EC2. I don't think EC2 will let you stall instance termination for such a long time. It's hard to find any documentation on how long an EC2 instance can inhibit the shutdown but I did see [this](https://serverfault.com/questions/295065/ec2-maximum-shutdown-time-exceeded) saying typically 10 minutes is the max.

The other thing I saw while investigating this is that Cluster API machine deletion has a [pre-terminate hook](https://github.com/kubernetes-sigs/cluster-api/blob/dd3c2515c3edab63ab08b506c612932ecf3b08f8/docs/proposals/20200602-machine-deletion-phase-hooks.md#pre-terminate). It seems like it might be possible to implement evicting DaemonSet pods by making a custom Hook Implementing Controller (HIC). Is that the preferred way to implement something like this? If so I can close this feature request and look into making the HIC.

/kind feature

**Comment by kfox1111** [comment]
Kind of sounds like https://github.com/kubernetes/kubernetes/issues/75482 :/

**Comment by ailurarctos** [comment]
> Kind of sounds like [kubernetes/kubernetes#75482](https://github.com/kubernetes/kubernetes/issues/75482) :/

Yes, I think if https://github.com/kubernetes/kubernetes/issues/75482 were implemented it could potentially be used to implement this feature request.

**Comment by atiratree** [comment]
This feature might be eventually supported with Declarative Node Maintenance: https://github.com/kubernetes/enhancements/pull/4213

**Comment by sbueringer** [comment]
Do we know how cluster-autoscaler implemented this feature?

In general the DaemonSet controller will add a toleration for the Unschedulable taint to all DaemonSet Pods (https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations).

So while it's possible to evict DaemonSet Pods they will just immediately be re-created.

**Comment by chrischdi** [comment]
Took a quick look at autoscalers code.

For me it looks like they don't handle that the daemonset controller schedules a new pod.
They ignore that fact but seem to evict the running pods once and seem to have the race that:
* if it was successful it just continues deletion (not listing pods again, just the existing ones before evicting ds pods are gone)
* if it was not successful it tries again, with then maybe other pods
